File: njk-typescript-implementation/src/validators/topologyValidator.ts
/**
 * Topology-Level Validator
 *
 * Extended validation beyond JSON Schema:
 * - Circular dependency detection
 * - Cross-service field references
 * - Enablement expression validation
 * - Provider field consistency
 * - Port conflicts
 * - Secret references
 */

import { readFileSync } from 'fs';
import type { Topology, Service, ValidationResult, ValidationError } from '../types.js';

export class TopologyValidator {
  private topology: Topology;
  private errors: ValidationError[] = [];
  private warnings: ValidationError[] = [];

  constructor(topology: Topology) {
    this.topology = topology;
  }

  /**
   * Run all validation levels
   */
  validate(): ValidationResult {
    this.errors = [];
    this.warnings = [];

    this.validateFieldLevel();
    this.validateServiceLevel();
    this.validateTopologyLevel();

    return {
      valid: this.errors.length === 0,
      errors: this.errors,
      warnings: this.warnings
    };
  }

  /**
   * Validate from file
   */
  static validateFile(topologyPath: string): ValidationResult {
    const content = readFileSync(topologyPath, 'utf-8');
    const topology = JSON.parse(content);
    const validator = new TopologyValidator(topology);
    return validator.validate();
  }

  // =========================================================================
  // FIELD LEVEL VALIDATION
  // =========================================================================

  private validateFieldLevel(): void {
    const services = this.topology.topology.services;

    for (const [serviceName, service] of Object.entries(services)) {
      const properties = service.configuration?.properties || {};

      for (const [fieldName, fieldDef] of Object.entries(properties)) {
        // Check sensitive fields have secret references
        if (fieldDef['x-sensitive'] && !fieldDef['x-secret-ref']) {
          this.warnings.push({
            path: `${serviceName}.${fieldName}`,
            message: 'Sensitive field lacks x-secret-ref',
            severity: 'warning'
          });
        }

        // Check env var is defined if field is exposed
        const visibility = fieldDef['x-visibility'] || 'exposed';
        if (['exposed', 'advanced'].includes(visibility) && !fieldDef['x-env-var']) {
          this.warnings.push({
            path: `${serviceName}.${fieldName}`,
            message: 'Exposed field lacks x-env-var definition',
            severity: 'warning'
          });
        }
      }
    }
  }

  // =========================================================================
  // SERVICE LEVEL VALIDATION
  // =========================================================================

  private validateServiceLevel(): void {
    const services = this.topology.topology.services;

    for (const [serviceName, service] of Object.entries(services)) {
      this.validateServiceDependencies(serviceName, service);
      this.validateServiceHealthcheck(serviceName, service);
    }

    // Port and container name validation needs to check across all services
    this.validateServicePorts();
    this.validateContainerNames();
  }

  private validateServiceDependencies(serviceName: string, service: Service): void {
    const requires = service.infrastructure.requires || [];
    const services = this.topology.topology.services;

    for (const dep of requires) {
      if (!services[dep]) {
        this.errors.push({
          path: serviceName,
          message: `Requires non-existent service '${dep}'`,
          severity: 'error'
        });
      }
    }
  }

  private validateServicePorts(): void {
    const services = this.topology.topology.services;
    const portMap = new Map<number, string[]>();

    for (const [serviceName, service] of Object.entries(services)) {
      const published = service.infrastructure.published_port;
      if (published !== null && published !== undefined) {
        if (!portMap.has(published)) {
          portMap.set(published, []);
        }
        portMap.get(published)!.push(serviceName);
      }
    }

    // Report conflicts
    for (const [port, serviceNames] of portMap.entries()) {
      if (serviceNames.length > 1) {
        this.errors.push({
          path: 'topology',
          message: `Port conflict: Port ${port} is published by multiple services: ${serviceNames.join(', ')}`,
          severity: 'error'
        });
      }
    }
  }

  private validateContainerNames(): void {
    const services = this.topology.topology.services;
    const nameMap = new Map<string, string[]>();

    for (const [serviceName, service] of Object.entries(services)) {
      const containerName = service.infrastructure.container_name;
      if (containerName) {
        if (!nameMap.has(containerName)) {
          nameMap.set(containerName, []);
        }
        nameMap.get(containerName)!.push(serviceName);
      }
    }

    // Report conflicts
    for (const [containerName, serviceNames] of nameMap.entries()) {
      if (serviceNames.length > 1) {
        this.errors.push({
          path: 'topology',
          message: `Container name conflict: '${containerName}' is used by multiple services: ${serviceNames.join(', ')}`,
          severity: 'error'
        });
      }
    }
  }

  private validateServiceHealthcheck(serviceName: string, service: Service): void {
    const infra = service.infrastructure;
    const enabled = infra.enabled || (infra.enabled_by && infra.enabled_by.length > 0);

    if (enabled && !infra.healthcheck) {
      this.warnings.push({
        path: serviceName,
        message: 'Enabled service lacks healthcheck',
        severity: 'warning'
      });
    }
  }

  // =========================================================================
  // TOPOLOGY LEVEL VALIDATION
  // =========================================================================

  private validateTopologyLevel(): void {
    this.validateNoCircularDependencies();
    this.validateEnablementExpressions();
    this.validateFieldReferences();
    this.validateSecretReferences();
    this.validateProviderConsistency();
    this.validateServiceEnablementReferences();
  }

  private validateNoCircularDependencies(): void {
    const services = this.topology.topology.services;
    const visited = new Set<string>();
    const recStack = new Set<string>();

    const hasCycle = (serviceName: string, path: string[] = []): boolean => {
      visited.add(serviceName);
      recStack.add(serviceName);

      const service = services[serviceName];
      if (!service) return false;

      const requires = service.infrastructure.requires || [];

      for (const dep of requires) {
        if (!visited.has(dep)) {
          if (hasCycle(dep, [...path, serviceName])) {
            return true;
          }
        } else if (recStack.has(dep)) {
          const cycle = [...path, serviceName, dep].join(' -> ');
          this.errors.push({
            path: 'topology',
            message: `Circular dependency detected: ${cycle}`,
            severity: 'error'
          });
          return true;
        }
      }

      recStack.delete(serviceName);
      return false;
    };

    for (const serviceName of Object.keys(services)) {
      if (!visited.has(serviceName)) {
        hasCycle(serviceName);
      }
    }
  }

  private validateEnablementExpressions(): void {
    const services = this.topology.topology.services;

    for (const [serviceName, service] of Object.entries(services)) {
      const enabledBy = service.infrastructure.enabled_by || [];

      for (const expression of enabledBy) {
        this.validateExpression(serviceName, expression);
      }
    }
  }

  private validateExpression(serviceName: string, expression: string): void {
    // Pattern: service.configuration.FIELD == value
    const match = expression.match(/^(\w+)\.configuration\.(\w+)\s*(==|!=)\s*(.+)$/);

    if (!match) {
      this.errors.push({
        path: serviceName,
        message: `Invalid expression format '${expression}'`,
        severity: 'error'
      });
      return;
    }

    const [, refService, refField] = match;

    // Check service exists
    if (!this.topology.topology.services[refService]) {
      this.errors.push({
        path: serviceName,
        message: `Expression references non-existent service '${refService}'`,
        severity: 'error'
      });
      return;
    }

    // Check field exists
    const properties = this.topology.topology.services[refService].configuration?.properties || {};
    if (!properties[refField]) {
      this.errors.push({
        path: serviceName,
        message: `Expression references non-existent field '${refService}.${refField}'`,
        severity: 'error'
      });
    }
  }

  private validateFieldReferences(): void {
    const services = this.topology.topology.services;

    for (const [serviceName, service] of Object.entries(services)) {
      const properties = service.configuration?.properties || {};

      for (const [fieldName, fieldDef] of Object.entries(properties)) {
        const requires = fieldDef['x-requires-field'];
        if (requires) {
          this.validateFieldReference(serviceName, fieldName, requires);
        }
      }
    }
  }

  private validateFieldReference(serviceName: string, fieldName: string, reference: string): void {
    const parts = reference.split('.');
    if (parts.length < 2) {
      this.errors.push({
        path: `${serviceName}.${fieldName}`,
        message: `Invalid reference format '${reference}'`,
        severity: 'error'
      });
      return;
    }

    const refService = parts[0];

    // Check service exists
    if (!this.topology.topology.services[refService]) {
      this.errors.push({
        path: `${serviceName}.${fieldName}`,
        message: `References non-existent service '${refService}'`,
        severity: 'error'
      });
    }
  }

  private validateSecretReferences(): void {
    const services = this.topology.topology.services;
    const secrets = this.topology.secrets || {};

    for (const [serviceName, service] of Object.entries(services)) {
      const properties = service.configuration?.properties || {};

      for (const [fieldName, fieldDef] of Object.entries(properties)) {
        const secretRef = fieldDef['x-secret-ref'];
        if (secretRef) {
          this.validateSecretReference(serviceName, fieldName, secretRef, secrets);
        }
      }
    }
  }

  private validateSecretReference(
    serviceName: string,
    fieldName: string,
    reference: string,
    secrets: any
  ): void {
    const parts = reference.split('.');

    if (parts[0] !== 'secrets') {
      this.errors.push({
        path: `${serviceName}.${fieldName}`,
        message: `Secret reference must start with 'secrets.'`,
        severity: 'error'
      });
      return;
    }

    // Navigate secrets object
    let obj: any = secrets;
    for (const segment of parts.slice(1)) {
      if (typeof obj !== 'object' || !obj[segment]) {
        this.errors.push({
          path: `${serviceName}.${fieldName}`,
          message: `Secret '${reference}' does not exist`,
          severity: 'error'
        });
        return;
      }
      obj = obj[segment];
    }
  }

  private validateProviderConsistency(): void {
    const services = this.topology.topology.services;

    for (const [serviceName, service] of Object.entries(services)) {
      const properties = service.configuration?.properties || {};

      for (const [fieldName, fieldDef] of Object.entries(properties)) {
        const providerFields = fieldDef['x-provider-fields'];
        if (providerFields && typeof providerFields === 'object' && !Array.isArray(providerFields)) {
          // Validate that referenced provider fields exist
          for (const [provider, fields] of Object.entries(providerFields)) {
            for (const requiredField of fields) {
              if (!properties[requiredField]) {
                this.errors.push({
                  path: `${serviceName}.${fieldName}`,
                  message: `Provider '${provider}' requires non-existent field '${requiredField}'`,
                  severity: 'error'
                });
              }
            }
          }
        }
      }
    }
  }

  private validateServiceEnablementReferences(): void {
    const services = this.topology.topology.services;

    for (const [serviceName, service] of Object.entries(services)) {
      const properties = service.configuration?.properties || {};

      for (const [fieldName, fieldDef] of Object.entries(properties)) {
        // Check x-enables-services
        const enables = fieldDef['x-enables-services'] || [];
        for (const enabledService of enables) {
          if (!services[enabledService]) {
            this.errors.push({
              path: `${serviceName}.${fieldName}`,
              message: `Enables non-existent service '${enabledService}'`,
              severity: 'error'
            });
          }
        }

        // Check x-affects-services
        const affects = fieldDef['x-affects-services'] || {};
        for (const [, affectedService] of Object.entries(affects)) {
          if (affectedService && !services[affectedService]) {
            this.errors.push({
              path: `${serviceName}.${fieldName}`,
              message: `Affects non-existent service '${affectedService}'`,
              severity: 'error'
            });
          }
        }
      }
    }
  }

  /**
   * Print validation results
   */
  printResults(result: ValidationResult, verbose = false): void {
    if (result.errors.length > 0) {
      console.log('❌ TOPOLOGY VALIDATION FAILED\n');
      console.log('Errors:');
      for (const error of result.errors) {
        console.log(`  • ${error.path}: ${error.message}`);
      }
    }

    if (result.warnings.length > 0) {
      console.log('\n⚠️  WARNINGS:');
      for (const warning of result.warnings) {
        console.log(`  • ${warning.path}: ${warning.message}`);
      }
    }

    if (result.valid && result.warnings.length === 0) {
      console.log('✅ TOPOLOGY VALIDATION PASSED');
      console.log('All cross-service relationships are valid');
    } else if (result.valid) {
      console.log(`\n✅ TOPOLOGY VALIDATION PASSED with ${result.warnings.length} warnings`);
    }

    if (verbose && result.valid) {
      this.printTopologySummary();
    }
  }

  private printTopologySummary(): void {
    const services = this.topology.topology.services;
    const serviceCount = Object.keys(services).length;

    const enabled = Object.values(services).filter(s => s.infrastructure.enabled).length;
    const conditional = Object.values(services).filter(s => (s.infrastructure.enabled_by || []).length > 0).length;
    const published = Object.values(services).filter(s => s.infrastructure.published_port !== null && s.infrastructure.published_port !== undefined).length;

    console.log('\n📊 TOPOLOGY SUMMARY:');
    console.log(`  Services: ${serviceCount}`);
    console.log(`  Enabled unconditionally: ${enabled}`);
    console.log(`  Enabled conditionally: ${conditional}`);
    console.log(`  Services with published ports: ${published}`);
  }
}

File: njk-python-implementation/validators/topology_validator.py
#!/usr/bin/env python3
"""
Topology-Level Validator

Implements extended validation rules beyond JSON Schema:
- Circular dependency detection
- Cross-service field reference validation
- Enablement expression validation
- Provider field consistency
- Port conflict detection
- Secret reference validation
"""

import json
import sys
import re
from pathlib import Path
from typing import Dict, List, Set, Optional, Tuple, Any
from collections import defaultdict, deque


class TopologyValidator:
    """Validates topology at field, service, and topology levels"""

    def __init__(self, topology: Dict):
        """
        Initialize validator with topology

        Args:
            topology: Topology dictionary
        """
        self.topology = topology
        self.errors: List[str] = []
        self.warnings: List[str] = []
        self.services = topology.get('topology', {}).get('services', {})

    def validate(self) -> bool:
        """
        Run all validation levels

        Returns:
            True if no errors, False otherwise
        """
        self.errors = []
        self.warnings = []

        self.validate_field_level()
        self.validate_service_level()
        self.validate_topology_level()

        return len(self.errors) == 0

    # =========================================================================
    # FIELD LEVEL VALIDATION
    # =========================================================================

    def validate_field_level(self):
        """Validate individual configuration fields"""
        for service_name, service in self.services.items():
            config = service.get('configuration', {})
            properties = config.get('properties', {})

            for field_name, field_def in properties.items():
                self._validate_field(service_name, field_name, field_def)

    def _validate_field(self, service_name: str, field_name: str, field_def: Dict):
        """Validate single field definition"""
        # Check sensitive fields have secret references
        if field_def.get('x-sensitive', False):
            if not field_def.get('x-secret-ref'):
                self.warnings.append(
                    f"{service_name}.{field_name}: Sensitive field lacks x-secret-ref"
                )

        # Check env var is defined if field should be exposed
        visibility = field_def.get('x-visibility', 'exposed')
        if visibility in ['exposed', 'advanced'] and not field_def.get('x-env-var'):
            self.warnings.append(
                f"{service_name}.{field_name}: Exposed field lacks x-env-var definition"
            )

    # =========================================================================
    # SERVICE LEVEL VALIDATION
    # =========================================================================

    def validate_service_level(self):
        """Validate individual services"""
        for service_name, service in self.services.items():
            self._validate_service_dependencies(service_name, service)
            self._validate_service_healthcheck(service_name, service)

        # Port validation needs to check across all services
        self._validate_service_ports()
        self._validate_container_names()

    def _validate_service_dependencies(self, service_name: str, service: Dict):
        """Ensure all required services exist"""
        requires = service['infrastructure'].get('requires', [])

        for dep in requires:
            if dep not in self.services:
                self.errors.append(
                    f"{service_name}: Requires non-existent service '{dep}'"
                )

    def _validate_service_ports(self):
        """Check for port conflicts"""
        port_map: Dict[int, List[str]] = defaultdict(list)

        for service_name, service in self.services.items():
            published = service['infrastructure'].get('published_port')
            if published is not None:
                port_map[published].append(service_name)

        # Report conflicts
        for port, services in port_map.items():
            if len(services) > 1:
                self.errors.append(
                    f"Port conflict: Port {port} is published by multiple services: {', '.join(services)}"
                )

    def _validate_container_names(self):
        """Check for container name conflicts"""
        name_map: Dict[str, List[str]] = defaultdict(list)

        for service_name, service in self.services.items():
            container_name = service['infrastructure'].get('container_name')
            if container_name:
                name_map[container_name].append(service_name)

        # Report conflicts
        for container_name, services in name_map.items():
            if len(services) > 1:
                self.errors.append(
                    f"Container name conflict: '{container_name}' is used by multiple services: {', '.join(services)}"
                )

    def _validate_service_healthcheck(self, service_name: str, service: Dict):
        """Warn if enabled service lacks healthcheck"""
        infra = service['infrastructure']
        enabled = infra.get('enabled', False) or len(infra.get('enabled_by', [])) > 0

        if enabled and not infra.get('healthcheck'):
            self.warnings.append(
                f"{service_name}: Enabled service lacks healthcheck"
            )

    # =========================================================================
    # TOPOLOGY LEVEL VALIDATION
    # =========================================================================

    def validate_topology_level(self):
        """Validate cross-service relationships"""
        self._validate_no_circular_dependencies()
        self._validate_enablement_expressions()
        self._validate_field_references()
        self._validate_secret_references()
        self._validate_provider_consistency()
        self._validate_service_enablement_references()

    def _validate_no_circular_dependencies(self):
        """Detect circular dependencies using DFS"""
        visited: Set[str] = set()
        rec_stack: Set[str] = set()

        def has_cycle(service_name: str, path: List[str] = None) -> bool:
            if path is None:
                path = []

            visited.add(service_name)
            rec_stack.add(service_name)

            if service_name not in self.services:
                return False

            service = self.services[service_name]
            requires = service['infrastructure'].get('requires', [])

            for dep in requires:
                if dep not in visited:
                    if has_cycle(dep, path + [service_name]):
                        return True
                elif dep in rec_stack:
                    cycle = ' -> '.join(path + [service_name, dep])
                    self.errors.append(f"Circular dependency detected: {cycle}")
                    return True

            rec_stack.remove(service_name)
            return False

        for service in self.services:
            if service not in visited:
                has_cycle(service)

    def _validate_enablement_expressions(self):
        """Ensure enabled_by conditions reference valid fields"""
        for service_name, service in self.services.items():
            enabled_by = service['infrastructure'].get('enabled_by', [])

            for expression in enabled_by:
                self._validate_expression(service_name, expression)

    def _validate_expression(self, service_name: str, expression: str):
        """
        Validate expression references valid fields

        Supports formats:
        - service.configuration.FIELD == value
        - service.configuration.FIELD != value
        """
        try:
            # Parse expression
            # Pattern: service.configuration.FIELD operator value
            match = re.match(r'^(\w+)\.configuration\.(\w+)\s*(==|!=)\s*(.+)$', expression)

            if not match:
                self.errors.append(
                    f"{service_name}: Invalid expression format '{expression}'"
                )
                return

            ref_service, ref_field, operator, value = match.groups()

            # Check service exists
            if ref_service not in self.services:
                self.errors.append(
                    f"{service_name}: Expression references non-existent service '{ref_service}'"
                )
                return

            # Check field exists
            config = self.services[ref_service].get('configuration', {})
            properties = config.get('properties', {})

            if ref_field not in properties:
                self.errors.append(
                    f"{service_name}: Expression references non-existent field '{ref_service}.{ref_field}'"
                )

        except Exception as e:
            self.errors.append(
                f"{service_name}: Malformed expression '{expression}': {e}"
            )

    def _validate_field_references(self):
        """Validate x-requires-field references"""
        for service_name, service in self.services.items():
            config = service.get('configuration', {})
            properties = config.get('properties', {})

            for field_name, field_def in properties.items():
                requires = field_def.get('x-requires-field')
                if requires:
                    self._validate_field_reference(service_name, field_name, requires)

    def _validate_field_reference(self, service_name: str, field_name: str, reference: str):
        """Check that referenced field exists"""
        segments = reference.split('.')
        if len(segments) < 2:
            self.errors.append(
                f"{service_name}.{field_name}: Invalid reference format '{reference}'"
            )
            return

        ref_service = segments[0]

        # Check service exists
        if ref_service not in self.services:
            self.errors.append(
                f"{service_name}.{field_name}: References non-existent service '{ref_service}'"
            )
            return

        # Navigate to referenced field
        try:
            obj = self.services[ref_service]
            for seg in segments[1:]:
                obj = obj.get(seg, {})

            # Check if we reached a valid endpoint
            if not obj and segments[1] in ['infrastructure', 'configuration']:
                self.warnings.append(
                    f"{service_name}.{field_name}: Reference '{reference}' may not exist at runtime"
                )

        except (KeyError, AttributeError):
            self.warnings.append(
                f"{service_name}.{field_name}: Could not fully validate reference '{reference}'"
            )

    def _validate_secret_references(self):
        """Validate x-secret-ref points to existing secrets"""
        secrets = self.topology.get('secrets', {})

        for service_name, service in self.services.items():
            config = service.get('configuration', {})
            properties = config.get('properties', {})

            for field_name, field_def in properties.items():
                secret_ref = field_def.get('x-secret-ref')
                if secret_ref:
                    self._validate_secret_reference(
                        service_name, field_name, secret_ref, secrets
                    )

    def _validate_secret_reference(
        self, service_name: str, field_name: str, reference: str, secrets: Dict
    ):
        """Check that secret path exists"""
        # Parse: "secrets.api_keys.litellm_master"
        segments = reference.split('.')

        if segments[0] != 'secrets':
            self.errors.append(
                f"{service_name}.{field_name}: Secret reference must start with 'secrets.'"
            )
            return

        # Navigate secrets object
        obj = secrets
        for seg in segments[1:]:
            if not isinstance(obj, dict) or seg not in obj:
                self.errors.append(
                    f"{service_name}.{field_name}: Secret '{reference}' does not exist"
                )
                return
            obj = obj[seg]

    def _validate_provider_consistency(self):
        """Validate provider fields are consistent with x-provider-fields definitions"""
        for service_name, service in self.services.items():
            config = service.get('configuration', {})
            properties = config.get('properties', {})

            for field_name, field_def in properties.items():
                provider_fields = field_def.get('x-provider-fields')
                if provider_fields and isinstance(provider_fields, dict):
                    # Validate that referenced provider fields exist
                    for provider, fields in provider_fields.items():
                        for required_field in fields:
                            if required_field not in properties:
                                self.errors.append(
                                    f"{service_name}.{field_name}: Provider '{provider}' requires non-existent field '{required_field}'"
                                )

    def _validate_service_enablement_references(self):
        """Validate services referenced in x-enables-services and x-affects-services exist"""
        for service_name, service in self.services.items():
            config = service.get('configuration', {})
            properties = config.get('properties', {})

            for field_name, field_def in properties.items():
                # Check x-enables-services
                enables = field_def.get('x-enables-services', [])
                for enabled_service in enables:
                    if enabled_service not in self.services:
                        self.errors.append(
                            f"{service_name}.{field_name}: Enables non-existent service '{enabled_service}'"
                        )

                # Check x-affects-services
                affects = field_def.get('x-affects-services', {})
                if isinstance(affects, dict):
                    for provider, affected_service in affects.items():
                        if affected_service and affected_service not in self.services:
                            self.errors.append(
                                f"{service_name}.{field_name}: Affects non-existent service '{affected_service}'"
                            )

    # =========================================================================
    # UTILITY METHODS
    # =========================================================================

    def print_results(self, verbose: bool = False):
        """Print validation results"""
        if self.errors:
            print("❌ TOPOLOGY VALIDATION FAILED\n")
            print("Errors:")
            for error in self.errors:
                print(f"  • {error}")

        if self.warnings:
            print("\n⚠️  WARNINGS:")
            for warning in self.warnings:
                print(f"  • {warning}")

        if not self.errors and not self.warnings:
            print("✅ TOPOLOGY VALIDATION PASSED")
            print("All cross-service relationships are valid")

        elif not self.errors:
            print(f"\n✅ TOPOLOGY VALIDATION PASSED with {len(self.warnings)} warnings")

        if verbose and not self.errors:
            self._print_topology_summary()

    def _print_topology_summary(self):
        """Print summary of topology structure"""
        print("\n📊 TOPOLOGY SUMMARY:")
        print(f"  Services: {len(self.services)}")

        # Count enabled services
        enabled = sum(
            1 for s in self.services.values()
            if s['infrastructure'].get('enabled', False)
        )
        conditional = sum(
            1 for s in self.services.values()
            if s['infrastructure'].get('enabled_by', [])
        )

        print(f"  Enabled unconditionally: {enabled}")
        print(f"  Enabled conditionally: {conditional}")

        # Count published ports
        published = sum(
            1 for s in self.services.values()
            if s['infrastructure'].get('published_port') is not None
        )
        print(f"  Services with published ports: {published}")


def main():
    """CLI entry point"""
    import argparse

    parser = argparse.ArgumentParser(
        description="Validate topology cross-service relationships",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Validation Rules:
  • No circular dependencies
  • All service dependencies exist
  • All enablement expressions reference valid fields
  • All cross-service field references are valid
  • All secret references exist
  • No port or container name conflicts
  • Provider field consistency

Examples:
  %(prog)s topology.json
  %(prog)s topology.json --verbose
        """
    )

    parser.add_argument(
        'topology',
        help='Path to topology.json file to validate'
    )

    parser.add_argument(
        '--verbose', '-v',
        action='store_true',
        help='Show verbose output with topology summary'
    )

    args = parser.parse_args()

    # Load topology
    try:
        with open(args.topology, 'r') as f:
            topology = json.load(f)
    except FileNotFoundError:
        print(f"Error: Topology file not found: {args.topology}")
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in topology file: {e}")
        sys.exit(1)

    # Validate
    validator = TopologyValidator(topology)
    valid = validator.validate()

    # Print results
    validator.print_results(verbose=args.verbose)

    # Exit with appropriate code
    sys.exit(0 if valid else 1)


if __name__ == '__main__':
    main()
